best-practices/rspec.md

```diff
@@ -0,0 +1,30 @@
+## Rspec style guide
+
+- First #describe what you are doing
+- Then establish the #context
+- #it only expects one thing (within reason)
+- Prefer explicitness - #it #its #specify #subject are nice shortcuts but
+consider whether they are sacrificing readability
+- Having excessive setup (fixtures, database entities, doubles etc.) can cause false-positives. Thus, set up preconditions as close to the test as possible (usually within the context block) and tear them down afterwards. This is really important when working with time for example.
+- Unit tests are fast and effective to work with when doing TDD. These are isolated tests, typically
+for a single class, and should use stubs and spies via test doubles to replace anything that is
+external to the _unit_ being tested. Testing side-effects that occur elsewhere in the system
+is incorrect. Not being able to unit test something is typically a signal that something may be wrong
+with the design of the unit
+- Beware of testing implementation specifics. When this becomes difficult, this typically identifies strong coupling - sometimes there is no way around this (e.g. you are calling an external API)
+- Avoid testing things that have already been tested elsewhere. Thin controller actions typically fall
+into this category, as do things like associations (e.g. something belongs_to some other thing)
+- Integration tests seek to test how various units integrate and thus typically involve assertions
+regarding side-effects. These tests can involve subsystems or the entire system - in the latter case, they are called feature tests. These types of tests are generally slow and should be reserved for the
+purpose of smoke testing key features of the site
```

Comment:
I've harped on this a few times, but I think this is a good chance to note it - using factories, then testing something with those factories, *are integration tests*; they test from where you start down the full stack to the DB, and are slow, both to setup, and to run.

The same effects can usually be tested by setting expectations on the models (e.g. `expect(teacher).to receive(:update)`), or (more usually,) something that uses the models.